Automattic/jetpack
# Assign unique text domains to our packages

#### Changes proposed in this Pull Request:

To finally fix #21690, we need each package to have its own textdomain and
to declare that domain in its `composer.json`.

The work done in previous PRs will take it from there: composer-plugin will
include the packages' domains in the generated `i18n-map.php` (#22096),
which the plugin will pass to the Assets package (#22143), which will map
those package textdomains to the plugin's (#22083).

This also adds checks to the monorepo linting script to help people get the
setup right:

* Packages may not use a textdomain that matches a plugin slug.
* If the package's phpcs config or eslint config sets a textdomain (which it
  will have to if it uses i18n after #22117 and #22131), require that same
  textdomain is set in composer.json.

#### Jetpack product discussion


Fixes #21690

#### Does this pull request change what data or activity we track or use?


No

#### Testing instructions:






* In the docker env or JN:
  * Build Jetpack or Backup (or use JN), and install (or generate) a language pack and set the site language to the pack.
    * Make sure only one of Jetpack, Backup, or Boost is activated to make things simpler.
  * Check that the plugin's `jetpack_vendor/i18n-map.php` contains mappings for all the packages.
  * Look in wp-admin => Tools, the entry for "Connection Manager" should be translated. This comes from the Connection package.
    * (If it's not, double check that your language pack includes a translation for that string).
  * Look at the output from /wp-json/jetpack/v4, in particular the descriptions under `.routes./jetpack/jetpack/v4/backup-helper-script.endpoints.0.args.helper.description` and `.routes./jetpack/jetpack/v4/backup-helper-script.endpoints.1.args.path.description`. These come from the Backup package.
    * (If they're not translated, double check that your language pack includes a translation for those strings).
* This probably needs some testing in wpcom too, but I'm not sure of specific messages to look at.
  * (???) Figure out some way to see messages that come from packages.
    * If nothing else, I suppose you could throw something like `_e( 'Connection Manager', 'jetpack-connection' )` at the end of `index.php`.
  * Install (or hack up) a language pack that includes translations for those messages.
  * Download and unpack the build artifact from this PR in your sandbox.
  * Update the `jetpack-packages/production/composer.json` with a repositories entry pointing to the unpacked sources, change all the versions to `@dev`, and do a composer update.
  * Verify that the `i18n-map.php` has mappings for all the domains.
  * Check that the messages are still translated.
